Release notes for bevy_color (#1202)
bevyengine/bevy-website: release-content/0.14/release-notes/12013_Upstreaming_bevy_color.md

Colors are a huge part of building a good game: UI, effects, shaders and more all need fully-feature, correct and convenient color tools.

> Bevy now supports a wide range of color spaces, each with their own type (e.g. `LinearRgba`, `Hsla` or `Oklaba`),

Reviewer: ```suggestion
Bevy now supports a broad selection of color spaces, each with their own type (e.g. `LinearRgba`, `Hsla` or `Oklaba`),
```

Nit: Mild rewording because "_wide range_" is used again at a latter point in the sentence.

> Bevy now supports a wide range of color spaces, each with their own type (e.g. `LinearRgba`, `Hsla` or `Oklaba`),

Reviewer 2: Agreed, although I prefer the Oxford comma ("A, B, and C").

and offers a wide range of fully documented operations on and conversions between them.

> Critically, color mixing / blending is now supported: perfect for procedurally generating color palettes and working with animations.

Reviewer 2: I don't know if we want to mention this, but the bevy_color can offer better performance in some cases, and is more idiomatic Rust.

Reviewer: Better performance I think is a side effect to the real win here in my opinion, which is we're able to give the user more control and transparency over how colours are represented. If a Bevy API wants a `LinearRgba` colour, now the user knows that and can adjust their algorithms accordingly (if they want to!).

Reviewer 2: Let's just say that I think we should do a better job of listing the benefits of bevy_color; after all, we're asking the users to do a fairly significant migration, so we ought to "sell" them on the advantages.

```rust

> use bevy_color::{palettes::tailwind, Mix};

Reviewer 3: I noticed that in this example we are importing a pallete and a trait, but not the colorspaces (like Srgba, Oklcha... ), can this be a problem?

Author: That *is* correct, maybe we should just use a glob import here.

Author: ```suggestion
use bevy_color::*;
```

// Each color space now corresponds to a specific type
let red = Srgba::rgb(1., 0., 0.);

> // All non-standard color space conversions are done through the `LinearRgba` color space
> // to avoid a quadratic explosion of generated code.
> let red = Oklcha::from(LinearRgba::from(red));

Reviewer: ```suggestion
// All non-standard color space conversions are done through the shortest path between
// the source and target color spaces to avoid a quadratic explosion of generated code.
// This conversion...
let red = Oklcha::from(red);
// ...is implemented using
let red = Oklcha::from(Oklaba::from(LinearRgba::from(red)));
```

I want to show that a user doesn't need to worry about _how_ a color is actually converted, since we've already ensured every conversion pair is available and takes the shortest path using well known conversion algorithms.

// We've added the `tailwind` palette colors: perfect for quick-but-pretty prototyping!

> let blue = Oklcha = tailwind::BLUE_500;

Reviewer 4: ```suggestion
// Also, we've reorganized the existing color palettes to make them consistent with industry standards.
let blue = Oklcha = tailwind::BLUE_500;
```

(This is a slight modification of @viridia's original wording, but it has the same meaning.)

// The color space that you're mixing your colors in has a huge impact!

> // Consider using the scientifically-motivated `Oklcha` or `Oklaba` for a perceptually linear effect

Reviewer 4: ```suggestion
// Consider using the scientifically-motivated `Oklcha` or `Oklaba` for a perceptually linear effect.
```

> // Consider using the scientifically-motivated `Oklcha` or `Oklaba` for a perceptually linear effect

Reviewer: ```suggestion
// Consider using the scientifically-motivated `Oklcha` or `Oklaba` for a perceptually uniform effect
```

Nit: "_perceptually uniform_" is the more common term used to describe this property.

let purple =  red.mix(blue, 0.5);
```

> Most of the user-facing APIs still accept a colorspace-agnostic `Color` (which now wraps our color-space types),

Reviewer 3: Will this change in later releases? Like, instead of using a global enum, the regular Impl Into<Insert the colorspace>

Author: Nope, I experimented with using `LinearRgba` everywhere and it was miserable ergonomics and the perf didn't change.

Reviewer: Yeah I think the `Color` enum was really hard to work with when all the colour spaces had to co-mingle inside it with no way in the type system to split off a particular type ("I don't need to convert this because I know it's already in this space", etc.) Now that we have specific types for each color space, the enum actually becomes much nicer to work with, to the point where removing it would make things worse. Shows how important experimenting is, because that wasn't the outcome I was expecting!

while rendering internals use the physically-based `LinearRgba` type.

> `bevy_color` offers a solid, type-safe foundation but it's just getting started.

Reviewer 4: ```suggestion
`bevy_color` offers a solid, type-safe foundation, but it's just getting started.
```

"it's just getting started" is an independent clause, so it needs a comma to introduce it. :)

Critically, color mixing / blending is now supported: perfect for procedurally generating color palettes and working with animations.

```rust

> use bevy_color::*;

Reviewer: ```suggestion
use bevy_color::prelude::*;
```

Nit: `bevy_color::*` would work, but we have a prelude so we might as well show it off.
